### ISSUE TITLE
Rethink connector breaks loopback context.

### ISSUE THREAD

**Issue by svarogg** [fuwaneko/loopback-connector-rethinkdb#7]
For some reason when using loopback-connector-rethinkdb in conjunction with loopback context, the context does not follow the request throughout the callback chain.

**Steps for reproduction**:
1. Clone https://github.com/svarogg/loopback-rethink-context.
2. Configure some rethinkdb datasource in server/datasources.json. Make sure it contains the tables required for loopback to work.
3. `npm install && slc run`
4. Go to http://localhost:3000/explorer
5. Set some access token
6. Send any request(for instance GET all users).
7. See in console output: `Context in server.js:  null`

**Control**:
1. Change server/datasources.json to the following:
{
  "db": {
    "name": "db",
    "connector": "memory"
  }
}
2. Repeat steps 3-6 from reproduction
3. See in console output:
Context in server.js:  { name: 'loopback',
  active: { accessToken: undefined },
  _set: [ null, { accessToken: undefined } ],
  id:
   { create: [Function],
     flags: 15,
     before: [Function],
     after: [Function],
     error: [Function],
     uid: 1,
     data: null } }
Or similar.

Thanks for any assistance.

**Comment by fuwaneko** [comment]
I'm not sure what the problem is, but you can use process.context["loopback"] as a workaround. I've tried reading sources for loopback and context.active is null, thus getCurrentContext does not return context. And context.active comes from continuation-local-storage module. I've checked if access token is fetched from db, etc, etc, and it does, I don't see anything wrong there.

**Comment by svarogg** [comment]
Hi and thanks for your quick reply.

Using process.context["loopback"] is something I have considered, but unfortunately, it is no good as a workaround.
Notice that when you print out process.context["loopback"] what you get is
    { name: 'loopback',
      active: null,
      _set: [],
      id:
       { create: [Function],
         flags: 15,
         before: [Function],
         after: [Function],
         error: [Function],
         uid: 1,
         data: null } }

Active and _set are null/empty respectively. Which basically means there's no data in the context.
In the working example, on the other hand, active and _set contained the access token, and any other information that was added to loopback context.

**Comment by fuwaneko** [comment]
@svarogg yeah, I've no idea why active contains nothing. It seems that it just doesn't get populated at some point, like before() function isn't called or something.

**Comment by svarogg** [comment]
I have compared to the mongoDB container, and looks like I have got it.
It seems like they are wrapping all their calls by some `bindToCurrentDomain` function.

I went ahead and done the same in the rethinkdb driver and seems like things are working now.
All that was required was in ast.js/run(connection, options, callback) to surround the callback with util.bindToCurrentDomain, and add to util.js:
module.exports.bindToCurrentDomain = function(callback) {
  var domain = process.domain;
  if(domain == null || callback == null) {
    return callback;
  } else {
    return domain.bind(callback);
  }
}

I will submit them a pull request to fix the problem but meanwhile, would you want to fix that problem locally in the connector?
I believe this can be done if every place that calls .run calls it with the wrapped callback.

**Comment by svarogg** [comment]
P.S.
Here's the method in a more clear way:
https://gist.github.com/svarogg/1552cd4e9dd675494021

**Comment by fuwaneko** [comment]
@svarogg I see. Domains are node.js stdlib to execute IO operations in groups. I'm not sure where it has to be done: driver or connector. If RethinkDB guys won't do it then I certainly will. Keep me updated.

**Comment by svarogg** [comment]
@fuwaneko So it looks like the issue will not be fixed in the driver, since I could not reproduce it without the loopback connector, and nobody knows enough about nodejs domains to be able to understand the real source of the problem.

I will try to fire up a pull request today, but not sure if will have time for that until next week..

**Comment by fuwaneko** [comment]
@svarogg It seems this is a common problem with continuation-local-storage module: strongloop-community/loopback-connector-elastic-search#15 See also strongloop/loopback#1495 and strongloop/loopback#1423

**Comment by svarogg** [comment]
Looks like it is a problem in blubird - the promise library that rethink driver uses.  
It will be eventually fixed, probably in bluebird, if not than in rethinkdb driver.  

So for now I'm closing this issue.
